### PR TITLE
Joe Lauletta - UI Design Patterns Challenge

### PR DESCRIPTION
**Pull request by jlappdev**

### CODE REVIEW

File: MVP Module/Source/Feed Presentation/FeedPresenter.swift [hunk 1]
```diff
@@ -13,12 +13,20 @@ protocol FeedView {
 	func display(_ viewModel: FeedViewModel)
 }
 
+protocol FeedErrorView {
+	typealias ErrorMessage = String
+
+	func display(_ errorMessage: ErrorMessage)
+}
```
Comment (danillahtin) [inline MVP Module/Source/Feed Presentation/FeedPresenter.swift:20]: The `message` could be an optional `String?`, so the Presenter can decide when to show/hide the message.

For example, in the `didStartLoadingFeed()` method, the Presenter can call  `feedErrorView.display(.none)` to hide the error.

This way, the ViewController doesn't need to decide when to hide the error. It's easier to manage presentation state transitions when all changes are centralized in the presenter.

Also It can be a bit more convenient to have a `FeedErrorViewModel` struct to describe error view instead of a `typealias ErrorMessage = String`. It is more typesafe and it would be easier to add new parameter (say some error icon) to `FeedErrorView` in future as it would require only the change in `FeedErrorViewModel` and will not require change in `FeedErrorView` protocol.

```suggestion
struct FeedErrorViewModel {
	let message: String?
}

protocol FeedErrorView {
	func display(_ viewModel: FeedErrorViewModel)
}
```

File: MVP Module/Source/Feed UI/Controllers/FeedViewController.swift [hunk 2]
```diff
@@ -22,6 +24,7 @@ public final class FeedViewController: UITableViewController, UITableViewDataSou
 	}
 
 	@IBAction private func refresh() {
+		errorView?.hideMessage()
```
Comment (danillahtin) [inline MVP Module/Source/Feed UI/Controllers/FeedViewController.swift:27]: The Presenter should decide when to hide/show errors. See my other comment.

File: MVVM Module/Source/Feed Presentation/FeedViewModel.swift [hunk 3]
```diff
@@ -20,12 +21,15 @@ final class FeedViewModel {
...
 	var onFeedLoad: Observer<[FeedImage]>?
+	var onErrorThrown: ErrorObserver?
 
 	func loadFeed() {
 		onLoadingStateChange?(true)
 		feedLoader.load { [weak self] result in
 			if let feed = try? result.get() {
 				self?.onFeedLoad?(feed)
+			} else {
+				self?.onErrorThrown?()
```
Comment (danillahtin) [inline MVVM Module/Source/Feed Presentation/FeedViewModel.swift:32]: `onErrorThrown: ErrorObserver?` could be `onErrorStateChange: Observer<String?>?`, so the ViewModel decides when to show/hide errors (instead of the view controller) and which text to present. It's easier to manage presentation state transitions when all changes are centralized.

For example, in the `loadFeed()` method, the ViewModel can call  `onErrorStateChange(.none)` to hide the error before loading the feed.

This way, the ViewController doesn't need to decide when to hide the error. It's easier to manage presentation state transitions when all changes are centralized in the view model.

```suggestion
	var onErrorStateChange: Observer<String?>?

	func loadFeed() {
		onLoadingStateChange?(true)
		onErrorStateChange?(nil)
		feedLoader.load { [weak self] result in
			if let feed = try? result.get() {
				self?.onFeedLoad?(feed)
			} else {
				self?.onErrorStateChange?(Localized.Feed.loadError)
```

File: MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift [hunk 4]
```diff
@@ -21,6 +23,7 @@ public final class FeedViewController: UITableViewController, UITableViewDataSou
 
 	@IBAction private func refresh() {
 		viewModel?.loadFeed()
+		errorView?.hideMessage()
```
Comment (danillahtin) [inline MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift:26]: The ViewModel should decide when to hide/show errors. See my other comment.

File: MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift [hunk 5]
```diff
@@ -32,6 +35,10 @@ public final class FeedViewController: UITableViewController, UITableViewDataSou
 				self?.refreshControl?.endRefreshing()
 			}
 		}
+
+		viewModel?.onErrorThrown = { [weak self] in
+			self?.errorView?.show(message: Localized.Feed.loadError)
```
Comment (danillahtin) [inline MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift:40]: The ViewModel should decide which error text to display. See my other comment.

**Review by danillahtin — COMMENTED** [review]
Hi @jlappdev, you almost passed the challenge!

This is Danil, I am an instructor at Essential Developer.

Please, check my review comments and let me know once you're done so I can review it again. 

Thank you 🙂

File: MVVM Module/Tests/Feed UI/FeedUISnapshotTests.swift [hunk 6]
```diff
@@ -65,7 +65,7 @@ private class FeedLoaderStub: FeedLoader {
 
 private extension FeedViewController {
 	func display(errorMessage: String) {
-		fatalError("Must be implemented - follow the MVC solution as a guide")
+		errorView?.show(message: errorMessage)
```
[the remark this answers is not in the document]
Comment (jlappdev) [reply MVVM Module/Tests/Feed UI/FeedUISnapshotTests.swift]: @danillahtin I wanted to better understand the distinct between using the `FeedViewController` reference to errorView and its reference to viewModel. Is the viewModel property not considered an internal property of `FeedViewController`?

**Comment by jlappdev** [comment]
@danillahtin I have addressed all of your concerns. Please review the changes. Also, I did have one question for you concerning a comment above. Thanks!

File: MVVM Module/Tests/Feed UI/FeedUISnapshotTests.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (danillahtin) [reply MVVM Module/Tests/Feed UI/FeedUISnapshotTests.swift]: Ideally `FeedViewController.viewModel` should not be used as well. It is better to control the viewModel from the outside. But due to the tests configuration and lack of external reference to the viewModel we can choose the lesser evil and address the viewModel via `FeedViewController.viewModel`.

Ideal solution would be to create new viewModel and then control it:
```swift
func display(errorMessage: String) {
        let viewModel = FeedViewModel(feedLoader: FeedLoaderStub(.success([])))
        self.viewModel = viewModel

        viewModel.onErrorStateChange?(errorMessage)
}
```
But as you can see it takes 3 times more code to implement than not ideal one.

File: MVP Module/Source/Feed UI/Controllers/FeedViewController.swift [hunk 7]
```diff
@@ -9,6 +9,8 @@ protocol FeedViewControllerDelegate {
 }
 
 public final class FeedViewController: UITableViewController, UITableViewDataSourcePrefetching, FeedLoadingView {
+	@IBOutlet private var errorView: ErrorView?
```
Comment (danillahtin) [inline MVP Module/Source/Feed UI/Controllers/FeedViewController.swift:12]: I recommend making `errorView` getter public on the production side and referencing it instead of casting the `tableHeaderView` to `ErrorView` on the test side.

Ideally, the view controller should be decoupled from the view hierarchy in the storyboard.

This way, you can easily change the view hierarchy in the storyboard without breaking the code here.

`errorView` setter can be private so that clients of `FeedViewController` cannot change its value.

```suggestion
	@IBOutlet public private(set) var errorView: ErrorView?
```

File: MVP Module/Tests/Feed UI/Helpers/FeedViewController+TestHelpers.swift [hunk 8]
```diff
@@ -57,4 +57,16 @@ extension FeedViewController {
 	private var feedImagesSection: Int {
 		return 0
 	}
+
+	var errorView: ErrorView? {
+		tableView.tableHeaderView as? ErrorView
+	}
+
```
Comment (danillahtin) [inline MVP Module/Tests/Feed UI/Helpers/FeedViewController+TestHelpers.swift:64]: To follow up on my previous comment this getter can be removed

```suggestion
```

File: MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift [hunk 9]
```diff
@@ -5,6 +5,8 @@
 import UIKit
 
 public final class FeedViewController: UITableViewController, UITableViewDataSourcePrefetching {
+	@IBOutlet private var errorView: ErrorView?
```
Comment (danillahtin) [inline MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift:8]: Following the same logic as for MVP solution I recommend making `errorView` getter public and setter private.

```suggestion
	@IBOutlet public private(set) var errorView: ErrorView?
```

File: MVVM Module/Tests/Feed UI/Helpers/FeedViewController+TestHelpers.swift [hunk 10]
```diff
@@ -57,4 +57,16 @@ extension FeedViewController {
 	private var feedImagesSection: Int {
 		return 0
 	}
+
+	var errorView: ErrorView? {
+		tableView.tableHeaderView as? ErrorView
+	}
+
```
Comment (danillahtin) [inline MVVM Module/Tests/Feed UI/Helpers/FeedViewController+TestHelpers.swift:64]: To follow up on my previous comment this getter can be removed
```suggestion
```

**Review by danillahtin — COMMENTED** [review]
Hi @jlappdev, almost there!

Please, check my review comments and let me know once you're done so I can review it again. 

Thank you 🙂

**Comment by jlappdev** [comment]
@danillahtin I have reverted the previous commits I had made which made the errorView property private. Thanks!

**Review by danillahtin — COMMENTED** [review]
Hi @jlappdev,

Congratulations, you've passed the challenge 🥇

- All tests are passing ✅
- Your code is clean and handles edge cases ✅
- The UI and animations look and behave as expected ✅
- Light and Dark mode support ✅
- Added an extra language ✅

Great job!
